drivers/adc/adc_sam_afec.c

```diff
@@ -279,6 +341,50 @@ static int adc_sam_init(const struct device *dev)
 		      | AFEC_MR_ONE
 		      | AFEC_MR_USEQ_NUM_ORDER;
 
+#ifdef CONFIG_ADC_SAM_AFEC_DMA
+	__ASSERT(cfg->trigger_source != 0,
+		"When DMA is enabled software triggering is not supported");
+	__ASSERT(cfg->trigger_source <= 6,
+		"Invalid trigger source, valid range is 0 - 6");
+	/* Enable hardware trigger */
+	afec->AFEC_MR |= AFEC_MR_TRGEN_EN;
+	afec->AFEC_MR |= AFEC_MR_TRGSEL(cfg->trigger_source);
+
+	data->timer = NULL;
+
+	/* If a timer is the trigger source, enable it now */
+	if (cfg->trigger_source == 1 ||
+	    cfg->trigger_source == 2 ||
+	    cfg->trigger_source == 3) {
+		switch (cfg->tc) {
+		case 0:
+			data->timer = device_get_binding("TC0");
+			break;
+		case 1:
+			data->timer = device_get_binding("TC1");
+			break;
+		case 2:
+			data->timer = device_get_binding("TC2");
+			break;
+		case 3:
+			data->timer = device_get_binding("TC3");
+			break;
```

Comment:
I don't know the hardware, but this should not be merged as-is in my opinion.

It looks like the devicetree should be using phandle properties to point to the right timer instances instead. Then you can just use DEVICE_DT_GET to get a pointer to the timer at build time.

I am against any changes to a devicetree binding that imply device_get_binding calls like these. A lot of effort has gone into making device_get_binding unnecessary now that DEVICE_DT_GET can be used to fetch devices from devicetree nodes directly.

Reply:
@mbolivar-nordic I agree that is better and that is how I had it originally. The issue I ran into is the case where there is a hardware trigger but it is not a timer. 

Perhaps, instead I can add another KConfig definition that enable/disable all the timer code, which is dependant on DMA being enabled. That way there will be no issue with the case where no timer is provided in the device tree since that code path will be disabled by the KConfig.

The other way is just to require the timer with DMA and not support the other hardware triggers. I will only be using the timer so this is fine for me, but this is less flexible. 

Thoughts?

Reply:
Can you explain a bit more what's going on in hardware?

I'm not familiar with how ADC peripheral DMA bindings work, but this is pretty suspicious in general. I would have expected something like this:

https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/iio/adc/st%2Cstm32-adc.yaml#L495

@anangl do you have any general process for how to provide DMA support in the ADC subsystem?

Also cc'ing @erwango since he may be able to shed some light on if the STM32 binding is not relevant here

Reply:
@mbolivar-nordic 

Basically the DMA on these devices have different slots for different peripherals. You configure the DMA to use any of the channels and a specific slot that is specific to the peripheral the transfer will be to/from. Instead of assigning it in the device tree this value is determined by the driver based on which ADC it is (ADC0 or ADC1)

You have to manually chose a channel in the device tree because the driver does not currently have a mechanism for providing a free channel.

As for the actual transfer, there is no configuration required on the ADC side, an event will be generated that notifies the DMA there is a sample ready, so long as the correct slot is configured.

The timer is only used here to trigger the ADC conversion, which in-turn notifies the DMA there is a sample waiting to be transferred. There is configuration required for the timer to be used as the trigger source as you can see in the code below:
```
	if (cfg->periph_id == ID_AFEC0) {
		dma_cfg.dma_slot = XDMAC_CC_PERID_AFEC0_Val;
	} else if (cfg->periph_id == ID_AFEC1) {
		dma_cfg.dma_slot = XDMAC_CC_PERID_AFEC1_Val;
	} else {
		__ASSERT_MSG_INFO("AFEC has an invalid peripheral ID!");
		return -EINVAL;
	}
```

The other option is to control the timer outside of the driver. But I found in this case the samples are not always in the correct order because the trigger timing is different each time. When the timer is started/stopped by the driver it is consistent.

I did not implement a way to use DMA without a hardware trigger as it would sort of defeat the purpose and would require extra complexity.

Like I said above I could change the binding to take the phandle of the timer instance instead and add a KConfig that enable/disable the timer code if someone wants to use a different hardware trigger like PWM or the analog comparator.

Right now it looks like this:
```
&afec0
{
	status = "okay";
	compatible = "atmel,sam-afec";
	dma-channel = <0>;
	tc = <0>;
	trigger-source = <1>;
};


&tc0 {
	status = "okay";
	reg-cmr = <0xCC003>;
	reg-rc = <125>;
	channel = <0>;
};
```
But would change to something like this (timer config being the same)

```
&afec0
{
	status = "okay";
	compatible = "atmel,sam-afec";
	dma-channel = <0>;
	tc = <&tc0>;
	trigger-source = <1>;
};
```
There is only a single DMA device that I why I opted not to include it in the devicetree and only include the channel number. Is this what you were referring to?

Reply:
@mbolivar-nordic I have changed the way this is done to how I think you would like it. The binding now takes a phandle to the timer and a phandle + channel options to the dma. 

I looked at the sam [i2s](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2s/i2s_sam_ssc.c) driver for inspiration.